File: src/main/java/com/leet/medium/SubarraySumEqualsK.cpp
#include <vector>
#include <unordered_map>
using namespace std;

class Solution
{
public:
	int subarraySum(vector<int> &nums, int k)
	{
		vector<int> prefix(nums.size());
		int ans = 0;
		int running = 0;
		for (int i = 0; i < nums.size(); i++)
		{
			running += nums[i];
			prefix[i] = running;
			if (running == k)
				ans++;
		}

		unordered_map<int, int> freq;

		for (int i = prefix.size() - 1; i >= 0; i--)
		{
			int cmpl = k + prefix[i];
			if (freq.find(cmpl) != freq.end())
			{
				int cur = freq[cmpl];
				ans += cur;
			}
			if (freq.find(prefix[i]) == freq.end())
			{
				freq[prefix[i]] = 0;
			}
			freq[prefix[i]]++;
		}

		return ans;
	}
};

File: src/main/java/com/leet/medium/SubarraySumEqualsK.rs
use std::collections::HashMap;

/**
560. Subarray Sum Equals K
https://leetcode.com/problems/subarray-sum-equals-k/
Medium

Given an array of integers nums and an integer k, return the total number of subarrays whose sum
equals to k.

Example 1:
Input: nums = [1,1,1], k = 2
Output: 2

Example 2:
Input: nums = [1,2,3], k = 3
Output: 2

Constraints:
  1 <= nums.length <= 2 * 104
  -1000 <= nums[i] <= 1000
  -107 <= k <= 107
 */

impl Solution {
  pub fn subarray_sum(nums: Vec<i32>, k: i32) -> i32 {
    let mut map: HashMap<i32, i32> = HashMap::new();
    let (mut running, mut ans) = (0, 0);
    map.insert(0, 1); // the empty subarray
    for num in nums {
      running += num;
      ans += map.get(&(running - k)).unwrap_or(&0);
      *map.entry(running).or_insert(0) += 1;
    }
    ans
  }
}
